### ISSUE TITLE
Case-control studies with variable index dates

### ISSUE THREAD

**Issue by iaindillingham** [opensafely-core/cohort-extractor#671]
We'd like to implement case-control studies with variable index dates, which means adding functionality to cohort-extractor. This functionality would read a file that contained a column of patient IDs and a column of index dates, and import these columns into the database. These columns would then be used:

* to filter the population, retaining only those patients with a patient ID in the column of patient IDs;
* as the value of the `index_date` variable in functions that accept a date expression, such as [`with_these_clinical_events`](https://docs.opensafely.org/study-def-variables/#cohortextractor.patients.with_these_clinical_events).

A study definition that uses this functionality would look something like this:

```python
from cohortextractor import StudyDefinition, patients

CONTROLS = "output/matched_matches.csv"

study = StudyDefinition(
    index_date="2021-01-01",  # Unused, but required
    case_index_date=patients.with_value_from_file(
        CONTROLS,
        returning="case_index_date",
        type="date",
    ),
    population=patients.which_exist_in_file(CONTROLS),  # Expects a patient_id column
    # Non-matching variables follow.
    # ...
)
```

Let's step through the important features of this study definition:

1. `CONTROLS` is a path to a file that, within the context of this issue, contains a column of patient IDs and a column of index dates. The patient IDs are those of the controls; the index dates are those of the cases. The researcher would have generated this file using either a scripted or a reusable action (opensafely-core/matching/issues/43). Whilst this file will probably be a CSV, we should consider supporting other file types. This file may contain more columns than the patient ID and index date columns; the names of the columns are not defined; the types of the columns are not defined.
2. `patients.with_value_from_file` is a new function. It returns the values of the given `column` with the given `type` in the given file. The name of this function is still open to discussion (see below).
3. `patients.which_exist_in_file` is a new function. It returns a list of integers of either `0` or `1` and is used to filter the population, retaining only those patients that exist in the given file (the `1`s). Again, the name of this function is still open to discussion (see below).

## Implementation

Iain's leaving this section vague, because he doesn't have a solid understanding of cohort-extractor. However, it's anticipated that the columns would be imported into the database within a new, probably temporary, table. Subsequent tables would be joined to this table; the values of the column of index dates would be used instead of a constant value (e.g. `StudyDefinition(index_date="2020-01-01"`).

## Names of functions

The sentence "The name of this function is still open to discussion" is shorthand for "The *specificity* of this function is open to discussion". As @sebbacon has highlighted, there are pros and cons associated with more generic functions, as described by this issue, and more specific functions, such as:

```python
study = StudyDefinition(
    population=patients.controls(CONTROLS),
    index_date="case_index_date",
    # Non-matching variables follow.
    # ...
)
```

## Further reading

For the scoping discussion of case-control studies with variable index dates, see: #585.

**Comment by iaindillingham** [comment]
First add `patients.with_value_from_file` to the public API:
* *cohortextractor/patients.py*

Then to each backend/backend test:
* TPP
  * *cohortextractor/tpp_backend.py*
  * *tests/test_tpp_backend.py*
* EMIS
  * *cohortextractor/emis_backend.py*
  * *tests/test_emis_backend.py*
* Databricks
  * *databricks_backend.py*
  * *tests/test_databricks_backend.py*

Then add `patients.which_exist_in_file` to the same files.

Questions:

* How to check the SQL generated by each backend?
* How to check that the patient ID/index date file (e.g. *matched_matches.csv*) is uploaded correctly?

**Comment by iaindillingham** [comment]
Closed with opensafely-core/cohort-extractor#694
